File: CP/TCS-Codevita/CodeVita2017-master/Round1/bombdrop.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct   // Point data type
{
    double x;
    double y;
}
Point;

typedef struct   // Line data type
{
    Point p1;
    Point p2;
}
Line;

// Emulate C++ constructors for structures
#define Point(x, y) ((Point) { x, y })
#define Line(p1, p2) ((Line) { p1, p2 })

// Distance between points
double pdistance(Point p1, Point p2)
{
    double dx, dy;

    dx = (p1.x - p2.x);
    dy = (p1.y - p2.y);

    return sqrt((dx * dx) + (dy * dy));
}

// Test point circle intersection
bool intersects(Point point, int x, int y, int r)
{
    return pdistance(point, Point(x, y)) <= r;
}

// Intersection distance, how much has the point penetrated into circle
double idistance(Point point, int x, int y, int r)
{
    return ((double) r) - pdistance(point, Point(x, y));
}

// Length of the line
double llength(Line line)
{
    return pdistance(line.p1, line.p2);
}

// Process a line, clears the area inside the circle, and returns the new length
double lprocess(Line line, int x, int y, int r)
{
    double distance;

    bool p1intersects;
    bool p2intersects;

    p1intersects = intersects(line.p1, x, y, r);
    p2intersects = intersects(line.p2, x, y, r);

    // If none of the points are in the circle, then return the length of the line
    if (!p1intersects && !p2intersects)
        return llength(line);

    // If both the points intersect, then the line is completely erased, 0 is the length
    if (p1intersects && p2intersects)
        return 0;

    // Compute the length of the line
    distance = llength(line);

    if (p1intersects)
        distance -= idistance(line.p1, x, y, r);

    else if (p2intersects)
        distance -= idistance(line.p2, x, y, r);

    return distance;
}

int main()
{
    int n, r, x, y, i, t1, t2, t3, t4;
    double count;

    scanf("%d\n%d\n%d %d", &n, &r, &x, &y);
    Line* lines = (Line*) malloc(sizeof(Line) * n);

    for (i = 0; i < n; i++)
    {
        scanf("%d %d %d %d", &t1, &t2, &t3, &t4);
        lines[i] = Line(Point(t1, t2), Point(t3, t4));
    }

    count = 0;

    for (i = 0; i < n; i++)
        count += lprocess(lines[i], x, y, r);

    printf("%.2lf\n", count);

    free (lines);
    return 0;
}

File: CP/TCS-Codevita/CodeVita2017-master/Round1/BombDrop.java
import java.util.Scanner;

public class BombDrop
{
    private static class Point
    {
        float x;
        float y;

        Point(float x, float y)
        {
            this.x = x;
            this.y = y;
        }

        float distance(Point p2)
        {
            float dx, dy;

            dx = (x - p2.x);
            dy = (y - p2.y);

            return (float) Math.sqrt((dx * dx) + (dy * dy));
        }
    }

    private static class Line
    {
        Point p1;
        Point p2;

        Line(Point p1, Point p2)
        {
            this.p1 = p1;
            this.p2 = p2;
        }

        float length()
        {
            return p1.distance(p2);
        }
    }

    private static class Circle
    {
        float r;
        Point center;

        Circle(int r, int x, int y)
        {
            this.r = r;
            this.center = new Point(x, y);
        }

        boolean intersects(Point point)
        {
            return point.distance(center) <= r;
        }

        float intersectionDistance(Point point)
        {
            return r - point.distance(center);
        }
    }

    private static float processLine(Line line, Circle circle)
    {
        boolean p1intersects = circle.intersects(line.p1);
        boolean p2intersects = circle.intersects(line.p2);

        if (!p1intersects && !p2intersects)
            return line.length();

        if (p1intersects && p2intersects)
            return 0;

        float distance = line.length();

        if (p1intersects)
            distance -= circle.intersectionDistance(line.p1);

        else if (p2intersects)
            distance -= circle.intersectionDistance(line.p2);

        return distance;
    }

    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();

        Circle circle = new Circle(sc.nextInt(), sc.nextInt(), sc.nextInt());
        Line[] lines = new Line[n];

        for (int i = 0; i < n; i++)
        {
            Point p1 = new Point(sc.nextInt(), sc.nextInt());
            Point p2 = new Point(sc.nextInt(), sc.nextInt());

            lines[i] = new Line(p1, p2);
        }

        float count = 0;

        for (int i = 0; i < n; i++)
            count += processLine(lines[i], circle);

        System.out.printf("%.2f\n", count);

        sc.close();
    }
}
